pkg/southbound/synchronizer/sessionManager.go

```diff
@@ -0,0 +1,279 @@
+// WithSessions sets list of sessions
+func WithSessions(sessions map[topodevice.ID]*Session) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.sessions = sessions
+	}
+}
+
+// WithDeviceStore sets device store
+func WithDeviceStore(deviceStore devicestore.Store) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.deviceStore = deviceStore
+	}
+}
+
+// WithMastershipStore sets mastership store
+func WithMastershipStore(mastershipStore mastership.Store) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.mastershipStore = mastershipStore
+	}
+}
+
+// WithOpStateChannel sets opStateChannel
+func WithOpStateChannel(opStateChan chan<- events.OperationalStateEvent) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.opStateChan = opStateChan
+	}
+}
+
+// WithSouthboundErrChan sets southbound error channel
+func WithSouthboundErrChan(southboundErrorChan chan events.DeviceResponse) func(*SessionManager) {
+	return func(factory *SessionManager) {
+		factory.southboundErrorChan = southboundErrorChan
+	}
+}
+
+// WithDispatcher sets dispatcher
+func WithDispatcher(dispatcher *dispatcher.Dispatcher) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.dispatcher = dispatcher
+	}
+}
+
+// WithModelRegistry sets model registry
+func WithModelRegistry(modelRegistry *modelregistry.ModelRegistry) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.modelRegistry = modelRegistry
+	}
+}
+
+// WithOperationalStateCache sets operational state cache
+func WithOperationalStateCache(operationalStateCache map[topodevice.ID]devicechange.TypedValueMap) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.operationalStateCache = operationalStateCache
+	}
+}
+
+// WithNewTargetFn sets southbound target function
+func WithNewTargetFn(newTargetFn func() southbound.TargetIf) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.newTargetFn = newTargetFn
+	}
+}
+
+// WithOperationalStateCacheLock sets operational state cache lock
+func WithOperationalStateCacheLock(operationalStateCacheLock *sync.RWMutex) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.operationalStateCacheLock = operationalStateCacheLock
+	}
+}
+
+// WithDeviceChangeStore sets device change store
+func WithDeviceChangeStore(deviceChangeStore device.Store) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.deviceChangeStore = deviceChangeStore
+	}
+}
+
+// Start starts session manager
+func (sm *SessionManager) Start() error {
+	log.Info("Session manager started")
+	err := sm.deviceStore.Watch(sm.topoChannel)
+	if err != nil {
+		return err
+	}
+
+	go sm.processDeviceEvents(sm.topoChannel)
+	go func() {
+		err := sm.updateDeviceState()
```

Comment:
I think we should create a separate event channel and goroutine for each session to allow session state changes to be notified and the topology to be updated concurrently. A new goroutine to process session state changes could be created in the session manager or in the session itself. We only need to maintain the order of updates within each session. Sharing a channel across sessions can only block the processing of session updates unnecessarily. A per-session loop could also simplify the algorithm for updating the device state.